### DISCUSSION TITLE
Middle Mouse Button Handling

### DISCUSSION THREAD

**Discussion by WerySkok** [flatpak/xdg-desktop-portal#1170] [New Portals]
Hello, I would like to request a separate interface for the handling of the middle mouse button behavior. Specifically, I would like to propose the following options:

1. **Paste from Selection Clipboard**: This option would mimic the traditional behavior from X, where clicking the MMB in an application would paste the content from the selection clipboard into that application. Thanks to Wayland transition, this option is now isn't the only possible one.

2. **Autoscrolling**: This option would enable autoscrolling, like in Windows. Clicking the MMB in a scrollable area would activate autoscrolling, allowing users to smoothly scroll through content without manually dragging the scrollbar.

3. **None**: This option would disable any specific middle mouse button behavior in xdg-desktop-portals, allowing applications to define their own custom functionality if desired.

**Comment by WerySkok** [comment]
I also think that enabling both Paste&Scroll should not be possible because that may lead to many undesired Paste actions in case if user just wanted to scroll through the content

**Reply by Mikenux** [reply]
What do you mean? That pasting and autoscrolling shouldn't happen at the same time when the MMB is pressed, or that pasting should be disabled when scrolling? I'm asking this because I don't know if it's possible to assign two shortcuts to the MMB.

**Reply by Play-InTheWay** [reply]
It probably means that pasting and autoscrolling shouldn't happen at the same time when the MMB is pressed. Something like that unfortunately can happen on Webcord and it's pretty undesirable.

**Reply by Mikenux** [reply]
Paste should be certainly disabled in both cases.

**Comment by Mikenux** [comment]
Is Autoscrolling implemented system-wide in a desktop environment?

**Reply by Play-InTheWay** [reply]
I don't think so, but it could be implemented from the toolkit or application side (as it kind of currently is).

I'm thinking that it may fit well in the Settings portal, though it wouldn't be an appearance setting.

**Reply by Mikenux** [reply]
> I don't think so, but it could be implemented from the toolkit or application side (as it kind of currently is).

Thanks. That's important as in a UI allowing the user to select which action to trigger from the MMB, we can't attest that it will autoscroll if it isn't sure it will.

> I'm thinking that it may fit well in the Settings portal, though it wouldn't be an appearance setting.

Don't know which portal it belongs to. In the settings portal and for the settings exposed, the user isn't asked for something. Here, I guess the app needs to declare if it supports MMB shortcuts and which ones. I also have https://github.com/flatpak/xdg-desktop-portal/discussions/1395, and this may be part of it (while not fully related) if it is approved.

**Comment by WerySkok** [comment]
KDE added autoscrolling in Plasma 6.2, though it works differently than on Windows: instead of acceleration based on the distance of the mouse from the starting point, Plasma just tied the scroll to the mouse position.

**Reply by Mikenux** [reply]
Ok, this is a libinput thing. The portal backend will indicate whether the DE supports the feature. Now I guess it's up to the portal maintainers to decide whether this would be part of a clipboard portal (thus conditioned on its development) or would be something independent.

**Reply by whot** [reply]
> Ok, this is a libinput thing. 

libinput has [on-button scrolling](https://wayland.freedesktop.org/libinput/doc/latest/scrolling.html#on-button-scrolling) which is a bit different to the autscrolling feature.

on-button scrolling unconditionally changes mouse motion events into scroll events while a given button is held down. autoscrolling usually works only on supported surfaces and has some related UI to indicate the actual scroll movements and the scroll itself isn't necessarily based on mouse movement but regular deltas generated based on the distance to the original point (like @WerySkok said).

The end result is roughly the same but it works on a higher level than the libinput feature.

**Reply by Mikenux** [reply]
Ok. From the moment event generation is system-side, that’s ok. Now, for the portal, is this something that should be an option? Or a call to mouse shortcut settings? (if we think there is no reason for an app to request autoscroll)

**Reply by Play-InTheWay** [reply]
I have tested the new autoscrolling of KDE. While it's ok for being a system-wide solution, it's more limited and uncomfortable than application/toolkit implementations and also conflicts with those, especially when in some apps, particularly the design ones, the handling is about dragging/moving parts of the workflow with the middle click.

Though, maybe with the portal in some way, the application/toolkit implementations could have priority over system-wide autoscrolling, so we can have the best of both worlds. But it would be possible or something that is worth working on? Keeping in mind that at the moment (and as I know), Plasma is the only one that supports something like that.

To some extent, I am more inclined to focus more on dealing with the issue between pasting with a middle click or autoscrolling in applications that support it instead of a system-wide implementation, as is on Windows.

**Reply by Mikenux** [reply]
For options, a backend can expose its own to the user, at least if the system has control over them. I think only options that are controlable by the portal itself should be exposed in the portal (i.e. paste, autoscroll, etc.).

Now,  the question to me is: Are the options applied for all apps or per app? What is expected?

**Reply by Play-InTheWay** [reply]
> Are the options applied for all apps or per app? What is expected?

That's why I said a system-wide (libinput autoscrolling) option would require extra work.

For the other options (enable/disable paste for selection or do nothing), I think being applied to all apps would be seen as an intuitive-expected behavior.

However, as I said, system-wide (libinput autoscrolling) conflicts with how apps currently use middle click, mainly applications that already implement autoscrolling like Firefox or whose handling is about dragging/moving parts of the workflow. So to implement this option, it must be through a per-app implementation.

**Reply by Mikenux** [reply]
After tought, wouldn't be preferable to have that per app if users prefers autoscrolling over paste (as there is the keyboard)? Also, I think we cannot show a window to choose a MMB shortcut applied system-wide (paste or app-side shortcut) as there is no paste for games. So, maybe it is preferable to go with a solution where there is a default, e.g. paste, and having apps requesting when they have other options than the default one. Another solution is just to let apps being able to have their own MMB shortcut, but only given we have something for paste (then, we will have to tell the user when the shortcut is paste). To see...
